File: app/src/main/java/com/example/androidscript/util/Interpreter/Interpreter.java
package com.example.androidscript.util.Interpreter;

import android.accessibilityservice.AccessibilityService;
import android.graphics.Bitmap;

import com.example.androidscript.util.AutoClick;
import com.example.androidscript.util.DebugMessage;
import com.example.androidscript.util.FileOperation;
import com.example.androidscript.util.ScreenShot;

import java.util.HashMap;
import java.util.Map;
import java.util.Vector;
import java.util.regex.Pattern;

public abstract class Interpreter extends Thread {//Every child only need to specify where and how to fetch files, as well as what kind of commands are accepted;

    protected abstract Vector<String> ReadCodeFromFile(String FileName);

    protected abstract Bitmap ReadImgFromFile(String FileName);

    public static class INVALID_CODE_EXCEPTION extends Exception {
        public INVALID_CODE_EXCEPTION(String s) {
            DebugMessage.set(s);
        }

        public INVALID_CODE_EXCEPTION() {
        }
    }

    public static class TargetImage {
        public Bitmap source;
        public int x1;//Starting point in whole graph
        public int y1;
        public int x2;
        public int y2;

        public TargetImage(Bitmap source, int x1, int y1, int x2, int y2) {
            this.source = source;
            this.x1 = x1;
            this.y1 = y1;
            this.x2 = x2;
            this.y2 = y2;
        }
    }

    public static class Code {//Every instance would be a valid code that interpreter can recognize
        public Vector<String[]> codes = new Vector<>();
        public Vector<String> dependency = new Vector<>();

        public Code(String[] SUPPORTED_COMMAND, Vector<String> RawCode) throws INVALID_CODE_EXCEPTION {
            for (String line : RawCode) {
                boolean valid = false;
                for (String format : SUPPORTED_COMMAND) {
                    if (Pattern.matches(format, line)) {
                        valid = true;
                        String[] command = line.split(" ");
                        if (command[0].equals("Call")) {
                            dependency.add(command[1]);
                        }
                        codes.add(command);
                        break;
                    }
                }
                if (!valid) {

                    throw new INVALID_CODE_EXCEPTION("Invalid " + line);
                }
            }
        }
    }


    protected Map<String, Interpreter.Code> MyCode = new HashMap<>();

    public abstract void Interpret(String FileName);

    protected void Interpret(String[] SUPPORTED_COMMAND, String FileName) throws Interpreter.INVALID_CODE_EXCEPTION {
        Vector<String> Command = ReadCodeFromFile(FileName);
        MyCode.put(FileName, new Code(SUPPORTED_COMMAND, Command));
        for (String depend : MyCode.get(FileName).dependency) {
            if (!MyCode.containsKey(depend)) {
                Interpret(SUPPORTED_COMMAND, depend);
            }
        }
    }

    //==========================================
    protected String run_arg_FileName = "";
    protected String[] run_arg_argv = null;
    protected int run_arg_depth = 0;
    protected AccessibilityService accessibilityService;

    public final void run(AccessibilityService accessibilityService, String FileName, String[] argv) throws RuntimeException {
        this.run_arg_FileName = FileName;
        this.run_arg_argv = argv;
        this.run_arg_depth = 0;
        this.accessibilityService = accessibilityService;
        this.start();
    }

    @Override
    public final void run() {
        if (this.run_arg_FileName.equals("")) {
            throw new RuntimeException("No code to run");
        }
        DebugMessage.set("RUN " + run_arg_FileName);
        this.run(run_arg_FileName, run_arg_argv, run_arg_depth);
    }

    protected void parseArguments(Map<String, String> LocalVar, String[] argv) {
        LocalVar.put("$R", "0");
        int argCount = 1;
        if (argv != null) {
            for (String arg : argv) {
                LocalVar.put("$" + argCount, arg);
                argCount++;
            }
        }
    }

    public int run(String FileName, String[] argv, int depth) throws RuntimeException {//Run code that is already read in MyCode
        assert (depth < 5);
        assert (MyCode.containsKey(FileName));
        Map<String, String> LocalVar = new HashMap<>();
        parseArguments(LocalVar, argv);

        int codeLength = MyCode.get(FileName).codes.size();
        for (int commandIndex = 0; commandIndex < codeLength; commandIndex++) {
            String[] command = (MyCode.get(FileName).codes.get(commandIndex));
            String[] Arguments = new String[command.length - 1];
            for (int i = 1; i < command.length; i++) {//Substitution
                if (command[i].charAt(0) == '$' && LocalVar.containsKey(command[i])) {//There might be Var command, that should replace $V
                    Arguments[i - 1] = LocalVar.get(command[i]);
                } else {
                    Arguments[i - 1] = command[i];
                }
            }
            sleep(300);
            switch (command[0]) {
                case "Click":
                    AutoClick.Click(accessibilityService, Integer.parseInt(Arguments[0]), Integer.parseInt(Arguments[1]));
                    LocalVar.put("$R", "0");
                    break;
                case "Compare":
                    if (ScreenShot.compare(FileOperation.instance.readPicAsBitmap(Arguments[4]), Integer.parseInt(Arguments[0]), Integer.parseInt(Arguments[1]), Integer.parseInt(Arguments[2]), Integer.parseInt(Arguments[3]), true)) {
                        LocalVar.put("$R", "0");
                    } else {
                        LocalVar.put("$R", "1");
                    }
                    break;
                case "JumpToLine":
                    commandIndex = Integer.parseInt(Arguments[0]) - 1;//One-based
                    LocalVar.put("$R", "0");
                    break;
                case "Wait":
                    sleep(Integer.parseInt(Arguments[0]));
                    LocalVar.put("$R", "0");
                    break;
                case "Call":
                    if (MyCode.containsKey(Arguments[0])) {
                        String[] nextArgv = new String[command.length - 2];
                        System.arraycopy(Arguments, 1, nextArgv, 0, command.length - 2);
                        int result = run(Arguments[0], nextArgv, depth + 1);
                        LocalVar.put("$R", String.valueOf(result));
                    } else {
                        LocalVar.put("$R", "1");
                    }
                    break;
                case "IfGreater":
                    DebugMessage.set(Arguments[0] + ":" + Integer.parseInt(Arguments[0]) + ">" + Arguments[1] + ":" + Integer.parseInt(Arguments[1]));
                    if (Integer.parseInt(Arguments[0]) <= Integer.parseInt(Arguments[1])) {//Failed, skip next line
                        commandIndex++;
                        LocalVar.put("$R", "1");
                    } else {
                        LocalVar.put("$R", "0");
                    }
                    break;
                case "IfSmaller":
                    if (Integer.parseInt(Arguments[0]) >= Integer.parseInt(Arguments[1])) {//Failed, skip next line
                        commandIndex++;
                        LocalVar.put("$R", "1");
                    } else {
                        LocalVar.put("$R", "0");
                    }
                    break;
                case "Var":
                    assert (Arguments[0].charAt(0) == '$');
                    LocalVar.put(Arguments[0], Arguments[1]);
                    LocalVar.put("$R", "0");
                    break;
                case "Exit":
                    Thread.currentThread().interrupt();
                    break;
                case "Return":
                    return Integer.parseInt(Arguments[0]);
                default:
                    throw new RuntimeException("Cannot Recognize " + command[0]);
            }

        }
        return 0;
    }

    protected void sleep(int ms) {
        try {
            Thread.sleep(ms);
        } catch (Exception e) {
            DebugMessage.printStackTrace(e);
        }
    }

    protected void delay(){
        sleep(500);
    }

}

File: app/src/main/java/com/example/androidscript/core/Interpreter.kt
package com.example.androidscript.core

import android.graphics.Bitmap
import android.util.Log
import com.example.androidscript.services.ClickService
import com.example.androidscript.services.WidgetService
import com.example.androidscript.services.ProjectionService
import com.example.androidscript.util.FileOperation
import com.example.androidscript.util.ImageHandler
import java.lang.Thread.sleep
import java.util.*

/** Every child needs to specify where and how to fetch files, as well as what kind of commands
 *  are accepted. */
open class Interpreter(
    private val ScriptFolderName: String,
    private val ScriptName: String,
    private val board: WidgetService.Bulletin
) : Runnable {

    private var scriptCode: HashMap<String, Code> = HashMap()
    var runningFlag = false

    init {
        interpret(ScriptName)
    }

    private fun readCode(fileName: String): Vector<String> {
        return FileOperation.readFromFileLines(ScriptFolderName + fileName)
    }

    private fun readImg(fileName: String): Bitmap {
        return FileOperation.readPicAsBitmap(ScriptFolderName + fileName)
    }

    private fun interpret(fileName: String) {
        val code = Code(readCode(fileName))

        val tagVar: HashMap<String, String> = HashMap()

        for (i in 0 until code.codes.size) {
            val command = code.codes[i]
            if (command[0] == "Tag") {
                tagVar[command[1]] = i.toString()
            }
        }

        for (line in code.codes) {
            for (i in 1 until line.size) { // The first word must be tag
                if (tagVar.containsKey(line[i])) {
                    line[i] = tagVar[line[i]].toString()
                }
            }
        }

        scriptCode[fileName] = code

        for (dependency in code.dependency) {
            if (!scriptCode.containsKey(dependency)) {
                interpret(dependency)
            }
        }
    }

    //==========================================
    private var runArgs: ArrayList<String>? = null
    private lateinit var screenShot: ProjectionService

    fun setup(argv: ArrayList<String>?, projectionService: ProjectionService) {
        runArgs = argv
        screenShot = projectionService
    }

    override fun run() {
        board.announce("Running")
        runningFlag = true

        try {
            execute(ScriptName, runArgs, 0)
            board.announce("IDLE")
        } catch (e: ClickService.AccessibilityServiceOffException) {
            Log.i(LOG_TAG, "AutoClickDead, terminating.")
            board.announce("AutoClick Failed.")
        } finally {
            runningFlag = false
        }
    }

    private fun execute(
        fileName: String,
        args: ArrayList<String>?,
        depth: Int,
    ): Int { //Run code that is already read in MyCode

        val localVar: MutableMap<String, String> = HashMap()
        parseArguments(localVar, args)

        var pc = 0

        while (runningFlag && (pc < scriptCode[fileName]!!.codes.size)) {
            val command = scriptCode[fileName]!!.codes[pc]
            val arguments = command.copyOfRange(1, command.size)
            varsSubstitution(localVar, command[0], arguments)
            //=====================================================
            Log.i(LOG_TAG, "$pc  ${command.joinToString(" ")}")
            delay()
            when (command[0]) {
                Commands.EXIT -> {
                    runningFlag = false
                    return 1
                }
                Commands.LOG -> board.announce(arguments[0])
                Commands.JUMP_TO -> pc = arguments[0].toInt() - 1 //One-based
                Commands.WAIT -> sleep(
                    arguments[0].toLong()
                )
                Commands.CALL -> localVar["\$R"] = execute(arguments[0], null, depth + 1).toString()
                Commands.TAG -> {}
                Commands.RETURN -> return arguments[0].toInt()
                Commands.CLICK_PIC -> {
                    val tmp = readImg(arguments[0])
                    val target = ImageHandler.findLocation(
                        screenShot.shot(), tmp, arguments[1]
                            .toDouble()
                    )
                    if (target != null) {
                        Log.i(LOG_TAG, "Clicking Picture:" + target.x + " " + target.y)
                        ClickService.click(target.x.toInt(), target.y.toInt())
                        localVar["\$R"] = "0"
                    } else {
                        localVar["\$R"] = "1"
                    }
                }
                Commands.CLICK -> {
                    ClickService.click(arguments[0].toInt(), arguments[1].toInt())
                }
                Commands.CALL_ARG -> {
                    val nextArgv =
                        arguments.copyOfRange(1, command.size - 1).toCollection(ArrayList())
                    localVar["\$R"] = execute(arguments[0], nextArgv, depth + 1).toString()
                }
                Commands.IF_GREATER -> if (arguments[0].toInt() <= arguments[1].toInt()) { //Failed, skip next line
                    pc++
                }
                Commands.IF_SMALLER -> if (arguments[0].toInt() >= arguments[1].toInt()) { //Failed, skip next line
                    pc++
                }
                Commands.ADD -> localVar[arguments[0]] = (localVar[arguments[0]]!!
                    .toInt() + arguments[1].toInt()).toString()
                Commands.SUBTRACT -> localVar[arguments[0]] = (localVar[arguments[0]]!!
                    .toInt() - arguments[1].toInt()).toString()
                Commands.VAR -> localVar[arguments[0]] = arguments[1]
                Commands.CHECK -> if (ImageHandler.checkColor(
                        screenShot.shot(), arguments[0]
                            .toInt(), arguments[1].toInt(), arguments[2].toInt()
                    )
                ) {
                    localVar["\$R"] = "0"
                } else {
                    localVar["\$R"] = "1"
                }
                Commands.SWIPE -> {
                    delay()
                    ClickService.swipe(
                        arguments[0].toInt(),
                        arguments[1].toInt(),
                        arguments[2].toInt(),
                        arguments[3].toInt(),
                    )
                }
                Commands.COMPARE -> {
                    screenShot.shot() //Empty shot to make sure Image be the newest.
                    localVar["\$R"] = screenShot.compare(
                        readImg(arguments[4]), arguments[0]
                            .toInt(), arguments[1].toInt(), arguments[2].toInt(), arguments[3]
                            .toInt()
                    ).toString() // similarity
                }
                else -> throw RuntimeException("Cannot Recognize " + command[0])
            }
            pc++
        }
        return 0
    }

    companion object {
        private val LOG_TAG = Interpreter::class.java.simpleName


        //==================== Helper =======================

        protected fun varsSubstitution(
            localVar: MutableMap<String, String>,
            command: String,
            args: Array<String>
        ) {
            if (args.isNotEmpty() && args[0][0] == '$' //Exit Has No Arg
                && command != "Var"
                && command != "Subtract"
                && command != "Add"
            ) {
                args[0] = localVar[args[0]]!!
            }
            for (z in 1 until args.size) {
                if (args[z][0] == '$') { //There might be Var command, that should replace $V
                    args[z] = localVar[args[z]]!!
                }
            }
        }

        protected fun parseArguments(
            LocalVar: MutableMap<String, String>,
            argv: ArrayList<String>?
        ) {
            LocalVar["\$R"] = "0"
            var argCount = 1
            argv?.run {
                for (arg in this) {
                    LocalVar["$$argCount"] = arg
                    argCount++
                }
            }
        }

        protected fun delay() {
            sleep(100)
        }
    }

    class InvalidCodeException(s: String) : Exception() {
        init {
            Log.d("InvalidCodeException", s)
        }
    }
}